cpp/src/io/orc/reader_impl.cu

```diff
@@ -413,33 +342,45 @@ rmm::device_buffer reader::impl::decompress_stripe_data(
         break;
       default: CUDF_FAIL("Unexpected decompression dispatch"); break;
     }
-    decompress_check(inflate_res, any_block_failure.device_ptr(), stream);
+
+    // Check if any block has been failed to decompress.
+    // Not using `thrust::any` or `thrust::count_if` to defer stream sync.
+    thrust::for_each(
+      rmm::exec_policy(stream),
+      thrust::make_counting_iterator(std::size_t{0}),
+      thrust::make_counting_iterator(inflate_res.size()),
+      [results           = inflate_res.begin(),
+       any_block_failure = any_block_failure.device_ptr()] __device__(auto const idx) {
+        if (results[idx].status != compression_status::SUCCESS) { *any_block_failure = true; }
+      });
```

Comment:
:clap:

Reply:
Reading the new code, it isn't clear that this does not care about synchronization; the old code had a comment indicating that this needn't be atomic. Could we pull that comment in here?

Reply:
Additionally, there might be value moving the definition of `any_block_failure` closer to this loop.

Reply:
Additionally, I wonder what the current state of `thrust::any_of()`'s performance is. Should we consider `thrust::count_if()` or another stand-in for `any_of()`? 

Unless I'm mistaken, all the kernel needs to return here is a single bool. The `host_device_vector[1]` seems unnecessary.

Reply:
This check is called multiple times for multiple data blocks but the output boolean value is only needed once after processing all blocks (which means it needs only 1 stream sync) so we can avoid multiple unnecessary stream sync.

Reply:
Let me see if I understand:

Why is `any_block_failure` a `host_device_vector` of size `1`? Why have an output variable at all? 

Why not use, say, `thrust::any_of()` to check `results[idx].status != SUCCESS`? (Because `any_of()` is slow?)

How about `count_if()`? Because that still requires a thread-sync before reducing the result?

Reply:
We need to do the check async for each block. Rather than sync after processing each block, we try to avoid sync altogether until we process all blocks. Then, we sync the check result and verify it on host.

`host_device_vector` of size `1` is just to store one boolean value (set to `false` if there is there is failure).

Both `any_of` and `count_if` sync stream after each block. As I said above, we don't want that.

Reply:
Ok, thanks for confirming. Makes sense.

I would still move the `any_block_failure` definition closer to its first use. I would also copy the original comment here:
```c++
// Doesn't need to be atomic
```

Reply:
Done.